File: packages/sdk/src/resources/balances.ts
import type { LinkOptions } from '@/config';
import { BaseResource } from '@/resources/base';
import type {
  IBalancesResource,
  ListBalancesParams,
} from '@/resources/interfaces';
import type { BalancesPage } from '@/types/index';
import { z } from 'zod';

const currencyAmountsSchema = z.record(z.string(), z.number());
const balanceSchema = z.looseObject({
  source_id: z.string(),
  type: z.enum(['cash', 'credit']),
  cash: z
    .looseObject({ available: currencyAmountsSchema })
    .nullable()
    .optional(),
  credit: z.looseObject({ used: currencyAmountsSchema }).nullable().optional(),
  current: z.number(),
  currency: z.string(),
  as_of: z.string(),
});
const balancesPageSchema = z.looseObject({
  data: z.array(balanceSchema),
  has_more: z.boolean().optional(),
});

export class BalancesResource
  extends BaseResource
  implements IBalancesResource
{
  constructor(options: LinkOptions) {
    super(options, '/balances');
  }

  private buildUrl(params: ListBalancesParams): string {
    const url = new URL(this.endpoint);

    if (params.sources !== undefined) {
      for (const source of params.sources) {
        url.searchParams.append('sources[]', source);
      }
    }
    if (params.limit !== undefined) {
      url.searchParams.set('limit', String(params.limit));
    }
    if (params.starting_after !== undefined) {
      url.searchParams.set('starting_after', params.starting_after);
    }
    if (params.ending_before !== undefined) {
      url.searchParams.set('ending_before', params.ending_before);
    }

    return url.toString();
  }

  async list(params: ListBalancesParams = {}): Promise<BalancesPage> {
    const { status, data, rawBody } = await this.apiFetch({
      method: 'GET',
      url: this.buildUrl(params),
    });

    if (status < 200 || status >= 300) {
      this.throwApiError('list balances', status, data, rawBody);
    }

    return this.parseResponse(
      'list balances',
      status,
      () => balancesPageSchema.parse(data) as BalancesPage,
    );
  }
}

File: packages/sdk-go/balances.go
package link

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

// BalancesResource provides balance operations.
type BalancesResource struct {
	base *baseResource
}

// List returns a page of balances.
func (r *BalancesResource) List(ctx context.Context, params *ListBalancesParams) (*BalancesPage, error) {
	endpoint, err := url.Parse(r.base.baseURL + "/balances")
	if err != nil {
		return nil, newConfigurationError("Invalid balances URL: " + err.Error())
	}
	query := endpoint.Query()
	if params != nil {
		for _, source := range params.Sources {
			query.Add("sources[]", source)
		}
		if params.Limit != nil {
			query.Set("limit", strconv.FormatInt(*params.Limit, 10))
		}
		if params.StartingAfter != nil {
			query.Set("starting_after", *params.StartingAfter)
		}
		if params.EndingBefore != nil {
			query.Set("ending_before", *params.EndingBefore)
		}
	}
	endpoint.RawQuery = query.Encode()
	response, err := r.base.fetch(ctx, http.MethodGet, endpoint.String(), nil, nil)
	if err != nil {
		return nil, err
	}
	if response.status < 200 || response.status >= 300 {
		return nil, newAPIError("list balances", response.status, response.data, response.rawBody)
	}
	var result BalancesPage
	if err := decodeResponse("list balances", response, &result); err != nil {
		return nil, err
	}
	return &result, nil
}
